planner/core/task.go

```diff
@@ -1813,17 +1836,102 @@ func (p *PhysicalHashAgg) attach2TaskForMpp(tasks ...task) task {
 		if !mpp.needEnforceExchanger(prop) {
 			return p.attach2TaskForMpp1Phase(mpp)
 		}
+		// we have to check it before the content of p has been modified
+		canUse3StageAgg := p.canUse3StageDistinctAgg()
 		proj := p.convertAvgForMPP()
 		partialAgg, finalAgg := p.newPartialAggregate(kv.TiFlash, true)
 		if finalAgg == nil {
 			return invalidTask
 		}
+
+		// generate 3 stage aggregation for single count distinct if applicable.
+		//  select count(distinct a), count(b) from foo
+		// will generate plan:
+		//  HashAgg sum(#1), sum(#2)                              -> final agg
+		//   +- Exchange Passthrough
+		//       +- HashAgg count(distinct a) #1, sum(#3) #2      -> middle agg
+		//           +- Exchange HashPartition by a
+		//               +- HashAgg count(b) #3, group by a       -> partial agg
+		//                   +- TableScan foo
+		var middleAgg *PhysicalHashAgg = nil
+		if partialAgg != nil && canUse3StageAgg {
+			clonedAgg, err := finalAgg.Clone()
+			if err != nil {
+				return invalidTask
+			}
+			middleAgg = clonedAgg.(*PhysicalHashAgg)
+			distinctPos := 0
+			middleSchema := expression.NewSchema()
+			schemaMap := make(map[int64]*expression.Column, len(middleAgg.AggFuncs))
+			for i, fun := range middleAgg.AggFuncs {
+				col := &expression.Column{
+					UniqueID: p.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  fun.RetTp,
+				}
+				if fun.HasDistinct {
+					distinctPos = i
+				} else {
+					fun.Mode = aggregation.Partial2Mode
+					originalCol := fun.Args[0].(*expression.Column)
+					schemaMap[originalCol.UniqueID] = col
+				}
+				middleSchema.Append(col)
+			}
+			middleAgg.schema = middleSchema
+
+			finalHashAgg := finalAgg.(*PhysicalHashAgg)
+			finalAggDescs := make([]*aggregation.AggFuncDesc, 0, len(finalHashAgg.AggFuncs))
+			for i, fun := range finalHashAgg.AggFuncs {
+				newArgs := make([]expression.Expression, 0, 1)
+				if distinctPos == i {
+					// change count(distinct) to sum()
+					fun.Name = ast.AggFuncSum
+					fun.HasDistinct = false
+					newArgs = append(newArgs, middleSchema.Columns[i])
+				} else {
+					for _, arg := range fun.Args {
+						newCol, err := arg.RemapColumn(schemaMap)
```

Comment:
We could use `ColumnSubstitute`?

Reply:
Ah, great. will do

Reply:
It turned out `ColumnSubstitute` is not applicable here, because the 2 schemas are different.